pkg/cloud/services/network/subnets.go

```diff
@@ -134,9 +134,21 @@ func (s *Service) reconcileSubnets() error {
 		existingSubnet := existing.FindEqual(sub)
 		if existingSubnet != nil {
 			subnetTags := sub.Tags
+
+			// TODO(mtulio): subnets need to be updated before skipping the loop (tag failure). Check edge cases.
+			// TODO(vincepri): check if subnet needs to be updated.
+			if len(sub.ID) > 0 {
+				// NOTE: Describing subnets assumes the subnet.ID is the same as the subnet's identifier (i.e. subnet-<xyz>),
+				// if we have a subnet ID specified in the spec, we need to restore it.
+				existingSubnet.ID = sub.ID
+			}
+
+			// Update subnet spec with the existing subnet details
+			existingSubnet.DeepCopyInto(sub)
```

Comment:
A dedicated PR has been created to isolate this change: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4917